madanalysis/layout/plotflow.py

```diff
@@ -719,157 +719,113 @@ def DrawMATPLOTLIB(
         for ind, hist in enumerate(histos):
             logging.getLogger("MA5").debug(f"<><><><><><> {hist.name} <><><><><><>")
 
-            # pdfset dictionary structure:
-            # "weights": are the nominal weights including scale uncertainties
-            # "replicas": are PDF variations
-            pdfset = {}
+            ########################################################################
+            # weight_set - dictionary structure:                                   #
+            #   - "weights": are nominal weights including scale uncertainties     #
+            #   - "pdf_variations": PDF variations                                 #
+            ########################################################################
+            weight_set = {}
             if len(datasets[ind].weight_collection) > 1:
-                # find pdf set
+                # separate the weights related to PDF variations from the other weights
                 for pdfid in datasets[ind].weight_collection.pdfsets:
                     if pdfid in self.pdftable:
-                        pdfset = copy.deepcopy(self.pdftable[pdfid])
-                        pdfset.update(
-                            {"weights": datasets[ind].weight_collection.pdfset(pdfid)}
-                        )
-                        pdfset.update({"replicas": WeightCollection()})
-                        for idx in range(1, pdfset["members"]):
-                            pdfset["replicas"] += datasets[ind].weight_collection.pdfset(
-                                pdfid + idx
-                            )
-                        break
+                        weight_set.update( {"weights": datasets[ind].weight_collection.pdfset(pdfid)} )
+                        weight_set.update( {"pdf_variations": {self.pdftable[pdfid]["name"]:WeightCollection()} } )
+                        for idx in range(1, self.pdftable[pdfid]["members"]):
+                            weight_set["pdf_variations"][self.pdftable[pdfid]["name"]] += datasets[ind].weight_collection.pdfset(pdfid + idx)
```

Comment:
setting the pdf name as an accessor here is not a good idea; then, we will have to track this within the code. This dictionary is not for users hence we don't need to know the name other than to lookup for details. ``pdfset`` dictionary did include all the information about the pdfset (see L730) ``"weights"`` are weight collection objects centralised around the main pdfset hence including scale variations. This is to handle scale uncertainties and to find nominal weight. ``"replicas"`` are different pdf replicas to be used to compute pdf uncertainties. Names of these objects are extra information, so setting dictionaries instead of lists will only make our lives harder. ``WeightCollection`` object is designed like a list with a searching functionality simply to achieve this.

Reply:
If the errors are computed from different PDF sets with independent ways to calculate the error, I don't see how to do it otherwise. What you propose only works is the PDF errors are evaluated with a single set. Any thoughts?